OSDOCS-7663: Added a new Quick Start Guide that uses Terraform (#64339)
openshift/openshift-docs: modules/rosa-sts-cluster-terraform.adoc

:tf-defaults:
endif::[]
:_content-type: PROCEDURE

= Creating your ROSA cluster with Terraform

The following Terraform cluster example creates your account-wide IAM roles then creates a ROSA cluster with a managed OIDC configuration.

.Procedure
. Optional: since the Terraform files that are created during this process are created in your current directory, you can create a new directory to store these files. Run the following command to create and navigate into a new directory:
+
[source,terminal]
----
$ mkdir terraform-cluster && cd terraform-cluster
----

. You can grant permissions to your account by using link:https://console.redhat.com/openshift/token[an offline {cluster-manager-first} token]. You set the token as a variable by running the following command:
+
[source,terminal]
----
$ export RHCS_TOKEN=<your_offline_token>
----
+
After you have exported your token, you can verify this value by running the following command:
+
[source,terminal]
----
$ echo $RHCS_TOKEN
----

. Create the Terraform files locally by using the following code templates:
.. Run the following command to create the `account-roles.tf` file:
+
[source,terminal]
----

> $ cat<<-EOF>account-roles.tf

Author: Step 4 (among others) seems to have it working, but good idea!

https://access.redhat.com/documentation/en-us/red_hat_openshift_service_on_aws/4/html/tutorials/cloud-experts-configure-custom-tls-ciphers

+
[source,terminal]
----
$ cat<<-EOF>account-roles.tf
data "rhcs_policies" "all_policies" {}

data "rhcs_versions" "all" {}

module "create_account_roles" {
  source  = "terraform-redhat/rosa-sts/aws"

>   version = "0.0.15"

Reviewer: ```suggestion
  version = ">=0.0.15"
```

  create_oidc_provider  = true

  cluster_id                  = ""
  rh_oidc_provider_thumbprint = rhcs_rosa_oidc_config.oidc_config.thumbprint
  rh_oidc_provider_url        = rhcs_rosa_oidc_config.oidc_config.oidc_endpoint_url
  tags                        = var.additional_tags
  path                        = var.path
}
EOF
----

.. Create the `operator-roles.tf` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>operator-roles.tf
module "operator_roles" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_operator_roles = true
  create_oidc_provider  = false

  rh_oidc_provider_thumbprint = rhcs_rosa_oidc_config.oidc_config.thumbprint
  rh_oidc_provider_url        = rhcs_rosa_oidc_config.oidc_config.oidc_endpoint_url
  operator_roles_properties   = data.rhcs_rosa_operator_roles.operator_roles.operator_iam_roles
  tags                        = var.additional_tags
  path                        = var.path
}
EOF
----

.. Create the `variables.tf` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>variables.tf
variable "rosa_openshift_version" {
  type        = string
  default     = "4.14.2"

>   description = "Desired version of OpenShift for the cluster, for example '4.1.0'. If version is greater than the currently running version, an upgrade will be scheduled."

Reviewer: ```suggestion
  description = "Desired version of OpenShift for the cluster, for example '4.14.2'. If version is greater than the currently running version, an upgrade will be scheduled."
```
Or any other , but bad example with non supported version)

    },
    operator_role_prefix = var.operator_role_prefix,
    oidc_config_id       = rhcs_rosa_oidc_config.oidc_config.id
  }
  worker_node_replicas = try(var.worker_node_replicas, var.multi_az ? 3 : 2)
}

data "aws_caller_identity" "current" {
}

resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster" {
  name                 = var.cluster_name
  cloud_region         = var.aws_region
  multi_az             = var.multi_az
  aws_account_id       = data.aws_caller_identity.current.account_id
  availability_zones   = var.availability_zones
  tags                 = var.additional_tags
  version              = var.rosa_openshift_version
  proxy                = var.proxy
  compute_machine_type = var.machine_type
  replicas             = local.worker_node_replicas
  autoscaling_enabled  = var.autoscaling_enabled
  min_replicas         = var.min_replicas
  max_replicas         = var.max_replicas
  sts                  = local.sts_roles
  properties = {
    rosa_creator_arn = data.aws_caller_identity.current.arn
  }
  #Private link settings
  private          = var.enable_private_link
  aws_private_link = var.enable_private_link

>   aws_subnet_ids   = var.create_vpc ? module.vpc[0].private_subnets : var.private_subnet_ids

Reviewer 2: `module.vpc` is not included in this example so this can simply be:

aws_subnet_ids = var.private_subnet_ids

>   machine_cidr     = var.enable_private_link ? var.vpc_cidr_block : null

Reviewer 2: This should be:
machine_cidr = var.vpc_cidr_block

>   depends_on = [time_sleep. wait_for_role_propagation]

Reviewer 2: Extra space here between time_sleep. and wait_for_role_propagation (my fault for typing too fast)

  description = "Enables autoscaling. This variable requires you to set a maximum and minimum replicas range using the 'max_replicas' and 'min_replicas' variables."
  type        = string
  default     = "false"
}

variable "min_replicas" {
  description = "The minimum number of replicas for autoscaling."
  type        = number
  default     = null
}

variable "max_replicas" {
  description = "The maximum number of replicas not exceeded by the autoscaling functionality."
  type        = number
  default     = null
}

variable "proxy" {
  default     = null
  description = "cluster-wide HTTP or HTTPS proxy settings"
  type = object({
    http_proxy              = string           # required  http proxy
    https_proxy             = string           # required  https proxy
    additional_trust_bundle = optional(string) # a string contains contains a PEM-encoded X.509 certificate bundle that will be added to the nodes' trusted certificate store.
    no_proxy                = optional(string) # no proxy
  })
}

#Private link cluster Info
variable "enable_private_link" {
  type        = bool
  description = "This enables private link"
  default     = false
}

#VPC Info

> variable "create_vpc" {

Reviewer 2: The code for this is not included in this example so this variable can be skipped

  type        = bool
  description = "Create custom VPC for ROSA cluster"
}

> variable "vpc_name" {

Reviewer 2: The code for this is not included in this example so this variable can be skipped

  type        = string
  description = "VPC Name"
  default     = null
}

variable "vpc_cidr_block" {
  type        = string
  description = "value of the CIDR block to use for the VPC"
  default     = "10.0.0.0/16"
}

> variable "private_subnet_cidrs" {

Reviewer 2: The code for this is not included in this example so this variable can be skipped

  type        = list(any)
  description = "The CIDR blocks to use for the private subnets"
  default     = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
}

> variable "public_subnet_cidrs" {

Reviewer 2: The code for this is not included in this example so this variable can be skipped

  type        = string
  default     = null
}

variable "private_subnet_ids" {
  type        = list(any)
  description = "VPC private subnets IDs for ROSA Cluster"
  default     = []
}
EOF
----

.. Run the following command to create your `vpc.tf` file:
+
[NOTE]
====
This guide does not use a VPC. You can change `create_vpc` to `true` in the `terraform.tfvars` file if you want to use a VPC.
====
+
[source,terminal]
----
$ cat<<-EOF>vpc.tf
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.1.2"

  count = var.create_vpc ? 1 : 0
  name  = var.vpc_name
  cidr  = var.vpc_cidr_block

  azs             = var.availability_zones
  private_subnets = var.private_subnet_cidrs
  public_subnets  = var.public_subnet_cidrs

  enable_nat_gateway   = true
  single_nat_gateway   = var.single_nat_gateway
  enable_dns_hostnames = true
  enable_dns_support   = true

  tags = var.additional_tags
}
EOF
----
. Create a `terraform.tfvars` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>terraform.tfvars
#AWS Info
aws_region="us-east-1"
availability_zones = ["us-east-1a"]

> redhat_aws_account_id="<AWS_account_ID>"

Reviewer 2: This can go

# Create a VPC?

> create_vpc="false"

Reviewer 2: If by default false, I would add a note saying that the user has to provide the correct `var. private_subnet_ids`, which is currently set to `[]` in the variables.tf

Reviewer 2: Either remove the default from variables.tf so the user needs to input it, or set create_vpc to true by default in my opinion

# Common ROSA Cluster Info

> cluster_name="rosa-tf-qsg"

Reviewer 2: I would make this something the user needs to type in personally

Author: I plan to do that with the next guide where I include customizations to the cluster. This is purely meant to be a pre-configured quick start.

# Advanced ROSA Cluster Options

> managed_oidc=true

Reviewer 2: Not needed as this example forces this to be true

.Procedure
. You can grant permissions to your account by using link:https://console.redhat.com/openshift/token[an offline {cluster-manager-first} token]. You set the token as a variable by running the following command:
+
[source,terminal]
----
$ export RHCS_TOKEN=<your_offline_token>
----
+
After you have exported your token, you can verify this value by running the following command:
+
[source,terminal]
----
$ echo $RHCS_TOKEN
----

. Create the Terraform files locally by using the following code templates:
+
[NOTE]
====
These files are created in your current directory. Ensure that you are in the directory where you want to run Terraform.
====

.. Run the following command to create your `account-roles.tf` file:
+
[source,terminal]
----
$ cat<<-EOF>account-roles.tf
data "rhcs_policies" "all_policies" {}

data "rhcs_versions" "all" {}

module "create_account_roles" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_account_roles  = true
  create_operator_roles = false

  account_role_prefix    = var.account_role_prefix
  path                   = var.path
  ocm_environment        = var.ocm_environment

>   rosa_openshift_version = regex("^[0-9]+\\\\.[0-9]+", var.rosa_openshift_version)

Reviewer 2: Extra \\ have been added in here somehow. This should only be two \\

Author: This renders as two `\`. The `\` is an escape symbol, so I needed 4 total.

Reviewer 2: Ok great!

  }
  worker_node_replicas = try(var.worker_node_replicas, var.multi_az ? 3 : 2)
}

data "aws_caller_identity" "current" {
}

resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster" {
  name                 = var.cluster_name
  cloud_region         = var.aws_region
  multi_az             = var.multi_az
  aws_account_id       = data.aws_caller_identity.current.account_id
  availability_zones   = var.availability_zones
  tags                 = var.additional_tags
  version              = var.rosa_openshift_version
  proxy                = var.proxy
  compute_machine_type = var.machine_type
  replicas             = local.worker_node_replicas
  autoscaling_enabled  = var.autoscaling_enabled
  min_replicas         = var.min_replicas
  max_replicas         = var.max_replicas
  sts                  = local.sts_roles
  properties = {
    rosa_creator_arn = data.aws_caller_identity.current.arn
  }
  #Private link settings
  private          = var.enable_private_link
  aws_private_link = var.enable_private_link
  aws_subnet_ids   = var.create_vpc ? module.vpc[0].private_subnets : var.private_subnet_ids
  machine_cidr     = var.enable_private_link ? var.vpc_cidr_block : null

>   depends_on = [module.create_account_roles]

Reviewer 2: This should be dependant on `time_sleep. wait_for_role_propagation`, not `module.create_account_roles`, otherwise it's skipping the sleep and trying to create too quickly: race condition

}

resource "rhcs_cluster_wait" "wait_for_cluster_build" {
  cluster = rhcs_cluster_rosa_classic.rosa_sts_cluster.id
  # timeout in minutes
  timeout = 60
}
EOF
----

.. Run the following command to create your `oidc-provider.tf` file:
+
[source,terminal]
----
$ cat<<-EOF>oidc-provider.tf
resource "rhcs_rosa_oidc_config" "oidc_config" {
  managed = true
}

data "rhcs_rosa_operator_roles" "operator_roles" {
  operator_role_prefix = var.operator_role_prefix
  account_role_prefix  = var.account_role_prefix
}

module "oidc_provider" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_operator_roles = false
  create_oidc_provider  = true

>   cluster_id                  = ""

Reviewer 2: Not used right now

  rh_oidc_provider_thumbprint = rhcs_rosa_oidc_config.oidc_config.thumbprint
  rh_oidc_provider_url        = rhcs_rosa_oidc_config.oidc_config.oidc_endpoint_url
  tags                        = var.additional_tags
  path                        = var.path
}
EOF
----

.. Run the following command to create your `operator-roles.tf` file:
+
[source,terminal]
----
$ cat<<-EOF>operator-roles.tf
module "operator_roles" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_operator_roles = true
  create_oidc_provider  = false

  rh_oidc_provider_thumbprint = rhcs_rosa_oidc_config.oidc_config.thumbprint
  rh_oidc_provider_url        = rhcs_rosa_oidc_config.oidc_config.oidc_endpoint_url
  operator_roles_properties   = data.rhcs_rosa_operator_roles.operator_roles.operator_iam_roles
  tags                        = var.additional_tags
  path                        = var.path
}
EOF
----

.. Run the following command to create your `variables.tf` file:
+
[source,terminal]
----
$ cat<<-EOF>variables.tf

> variable "redhat_aws_account_id" {

Reviewer 2: This variable is not needed

  type    = string
}

> variable "ocm_environment" {

Reviewer 2: This variable is not needed

  description = "account role policies details for account roles creation"
  type = object({
    sts_installer_permission_policy             = string
    sts_support_permission_policy               = string
    sts_instance_worker_permission_policy       = string
    sts_instance_controlplane_permission_policy = string
  })
  default = null
}

# Operator Roles
  variable "operator_role_prefix" {
  type    = string
  default = "ROSA"
}

variable "operator_role_policies" {
  description = "operator role policies details for operator roles creation"
  type = object({
    openshift_cloud_credential_operator_cloud_credential_operator_iam_ro_creds_policy = string
    openshift_cloud_network_config_controller_cloud_credentials_policy                = string
    openshift_cluster_csi_drivers_ebs_cloud_credentials_policy                        = string
    openshift_image_registry_installer_cloud_credentials_policy                       = string
    openshift_ingress_operator_cloud_credentials_policy                               = string
    openshift_machine_api_aws_cloud_credentials_policy                                = string
  })
  default = null
}

> variable "managed_oidc" {

Reviewer 2: As this example forces a managed OIDC provider, this variable can be removed for the sake of readability

+
[source,terminal]
----
$ echo $RHCS_TOKEN
----

. Create the Terraform files locally by using the following code templates:
+
[NOTE]
====
These files are created in your current directory. Ensure that you are in the directory where you want to run Terraform.
====

.. Run the following command to create your `account-roles.tf` file:
+
[source,terminal]
----
$ cat<<-EOF>account-roles.tf
data "rhcs_policies" "all_policies" {}

data "rhcs_versions" "all" {}

module "create_account_roles" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_account_roles  = true
  create_operator_roles = false

  account_role_prefix    = var.account_role_prefix
  path                   = var.path
  rosa_openshift_version = regex("^[0-9]+\\\\.[0-9]+", var.rosa_openshift_version)
  account_role_policies  = data.rhcs_policies.all_policies.account_role_policies
  all_versions           = data.rhcs_versions.all
  operator_role_policies = data.rhcs_policies.all_policies.operator_role_policies
  tags                   = var.additional_tags
}

resource "time_sleep" "wait_for_role_propagation" {
  depends_on = [ module.create_account_roles ]

  create_duration = "30s"
}
EOF
----

> .. Run the following command to create the `main.tf` file:

Reviewer 3: Nitpick, but this line says "the" (followed by command name). I believe the others say "your".

> .. Run the following command to create the `main.tf` file:

Reviewer 3: Also, given the note below, is this step optional?

  description = "value of the CIDR block to use for the VPC"
  default     = "10.0.0.0/16"
}

variable "private_subnet_cidrs" {
  type        = list(any)
  description = "The CIDR blocks to use for the private subnets"
}

variable "public_subnet_cidrs" {
  type        = list(any)
  description = "The CIDR blocks to use for the public subnets"
  default     = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]
}

variable "single_nat_gateway" {
  type        = bool
  description = "Single NAT or per NAT for subnet"
  default     = false
}

#AWS Info
variable "aws_region" {
  type    = string
  default = "us-east-1"
}

variable "availability_zones" {
  type        = list(any)
  description = "The availability zones to use for the cluster"
  default     = ["us-east-1a"]
}

variable "installer_role_arn" {
  description = "STS Role ARN with get secrets permission, relevant only for unmanaged OIDC config"
  type        = string
  default     = null
}

variable "private_subnet_ids" {
  type        = list(any)
  description = "VPC private subnets IDs for ROSA Cluster"
  default     = []
}
EOF
----

.. Run the following command to create your `vpc.tf` file:
+
[NOTE]
====

> This guide does not use a VPC. You can change `create_vpc` to `true` in the `terraform.tfvars` file if you want to use a VPC.

Reviewer 3: Given this note, is the step optional?

Author: It wouldn't be simply because if you don't have that VPC module, it fails. I want to add VPC in another guide, but I want to reuse this information. I could possibly create a conditionalized version without the VPC though.

====
+
[source,terminal]
----
$ cat<<-EOF>vpc.tf
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.1.2"

  count = var.create_vpc ? 1 : 0
  name  = var.vpc_name
  cidr  = var.vpc_cidr_block

  azs             = var.availability_zones
  private_subnets = var.private_subnet_cidrs
  public_subnets  = var.public_subnet_cidrs

  enable_nat_gateway   = true
  single_nat_gateway   = var.single_nat_gateway
  enable_dns_hostnames = true
  enable_dns_support   = true

  tags = var.additional_tags
}
EOF
----
. Create a `terraform.tfvars` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>terraform.tfvars
# Advanced ROSA Cluster Options
worker_node_replicas = "2"

#Private Link Cluster & VPC
single_nat_gateway=true
vpc_cidr_block="10.12.0.0/16"
private_subnet_cidrs=["10.12.0.0/22", "10.12.4.0/22", "10.12.8.0/22"]
public_subnet_cidrs=["10.12.128.0/24", "10.12.129.0/24", "10.12.130.0/24"]
EOF
----

> . In your terminal, run the following command in the directory where you saved these Terraform files to set up Terraform to create these resources:

Reviewer 3: ```suggestion
. In your terminal, run the following command in the directory where you saved the Terraform files to set up Terraform to create these resources:
```

Reviewer 3: One too many "these" for my liking:) TBH, would like to come up with another way to phrase "to set up" as well but I am coming up with nothing :)

+
[source,terminal]
----
$ terraform init
----

. Optional: Run the following command to confirm that the Terraform you copied is correct:
+
[source,terminal]
----
$ terraform validate
----
+

> You can also optionally run the following command to test the creation and create a reusable Terraform plan file:

Reviewer 3: Also, it looks like there is no command there, just the sample output?

> You can also optionally run the following command to test the creation and create a reusable Terraform plan file:

Author: I see what happened. I originally had the `terraform plan` command so you would have a .tfplan file after you run this. Speaking with a few folks, that doesn't seem to be that common, so I removed that action but left this step. Good catch!

+
.Sample output
+
[source,terminal]
----
Success! The configuration is valid.
----

. Run the following command to build your cluster with Terraform:
+
[source,terminal]
----
$ terraform apply
----

. When prompted, enter a name for your cluster.
+
[source,terminal]
----
var.cluster_name
  The name of the ROSA cluster to create

  Enter a value:
----
+

> After entering a cluster name, you see the Terraform plan, noting what is created or changed. Enter `yes` to start the role and cluster creation.

Reviewer 3: Not sure if we should say "you see" here?

Reviewer 3: Should "Enter 'yes' be a step?

Reviewer 2: It's quite clearly stated by the terraform command when you run it

+
[IMPORTANT]
====

> Your cluster name must be fewer than 16 characters, or you receive an error.

Reviewer 3: ```suggestion
Your cluster name must be fewer than 16 characters.
```

  description = "VPC private subnets IDs for ROSA Cluster"
  default     = []
}
EOF
----

.. Run the following command to create your `vpc.tf` file:
+
[NOTE]
====
This guide does not use a VPC. You can change `create_vpc` to `true` in the `terraform.tfvars` file if you want to use a VPC.
====
+
[source,terminal]
----
$ cat<<-EOF>vpc.tf
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.1.2"

  count = var.create_vpc ? 1 : 0
  name  = var.vpc_name
  cidr  = var.vpc_cidr_block

  azs             = var.availability_zones
  private_subnets = var.private_subnet_cidrs
  public_subnets  = var.public_subnet_cidrs

  enable_nat_gateway   = true
  single_nat_gateway   = var.single_nat_gateway
  enable_dns_hostnames = true
  enable_dns_support   = true

  tags = var.additional_tags
}
EOF
----
. Create a `terraform.tfvars` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>terraform.tfvars
#AWS Info
aws_region="us-east-1"
availability_zones = ["us-east-1a"]

# Create a VPC?
create_vpc="false"

# Advanced ROSA Cluster Options
rosa_openshift_version="4.13.17"
machine_type="m5.xlarge"
worker_node_replicas = "2"
multi_az = "false"

> #When this is enabled, you cannot configure the work_node_replicas

Reviewer 4: worker_node_replicas

Reviewer 4: 'When _____ is enabled'
Should be more specific.

autoscaling_enabled = "false"
#Change min to at least 3 for HA

> #If autoscaling_enabled is false, then these must be null

Reviewer 4: must we include the min/max params if they're null?
